src/main/java/com/amazonaws/encryptionsdk/keyrings/AwsKmsSymmetricMultiRegionDiscoveryKeyringBuilder.java

```diff
@@ -0,0 +1,153 @@
+        return this;
+    }
+
+    /**
+     * An optional ClientConfiguration for use with every AWS SDK KMS service client.
+     *
+     * @param clientConfiguration Custom ClientConfiguration to use.
+     * @return The AwsKmsSymmetricMultiRegionDiscoveryKeyringBuilder, for method chaining
+     */
+    public AwsKmsSymmetricMultiRegionDiscoveryKeyringBuilder clientConfiguration(ClientConfiguration clientConfiguration) {
+        this.clientConfiguration = clientConfiguration;
+        return this;
+    }
+
+    /**
+     * A list of string grant tokens to be included in all KMS calls.
+     *
+     * @param grantTokens The list of grant tokens.
+     * @return The AwsKmsSymmetricMultiRegionDiscoveryKeyringBuilder, for method chaining
+     */
+    public AwsKmsSymmetricMultiRegionDiscoveryKeyringBuilder grantTokens(List<String> grantTokens) {
+        this.grantTokens = grantTokens;
+        return this;
+    }
+
+    /**
+     * A list of AWS regions Ids identifying the AWS regions to attempt decryption in.
+     *
+     * @param regionIds The list of regions.
+     * @return The AwsKmsSymmetricMultiRegionDiscoveryKeyringBuilder, for method chaining
+     */
+    public AwsKmsSymmetricMultiRegionDiscoveryKeyringBuilder regions(List<String> regionIds) {
+        this.regionIds = regionIds;
+        return this;
+    }
+
+    /**
+     * An AWS Account Id to limit decryption to encrypted data keys for a specific AWS account.
+     *
+     * @param awsAccountId An AWS account id.
+     * @return The AwsKmsSymmetricMultiRegionDiscoveryKeyringBuilder, for method chaining
+     */
+    public AwsKmsSymmetricMultiRegionDiscoveryKeyringBuilder awsAccountId(String awsAccountId) {
+        this.awsAccountId = awsAccountId;
+        return this;
+    }
+
+    /**
+     * Constructs the {@code MultiKeyring} of {@code AwsKmsSymmetricRegionDiscoveryKeyring}s.
+     *
+     * @return The {@link Keyring} instance
+     */
+    public MultiKeyring build() {
+        // A mapping of AWS region to DataKeyEncryptionDao
+        final Map<String, DataKeyEncryptionDao> clientMapping = new HashMap<>();
+
+        // Construct each AwsKmsSymmetricRegionDiscoveryKeyring
+        List<Keyring> discoveryKeyrings = new ArrayList<>();
+        if (this.regionIds != null) {
+            for (final String region : this.regionIds) {
+                if (StringUtils.isBlank(region)) {
+                    continue;
+                }
+
+                // Check if a dao already exists for the given region
+                // and use the existing dao or construct a new one and save it
+                final boolean discoveryDaoExists = clientMapping.containsKey(region);
+                final DataKeyEncryptionDao discoveryDao = discoveryDaoExists ? clientMapping.get(region) : constructDataKeyEncryptionDao(region);
+                if (!discoveryDaoExists) {
+                    clientMapping.put(region, discoveryDao);
+                }
+
```

Comment:
Handling region duplication can run into a variety of implementation bugs. For example, in some languages, it might be easy to convert the list into a set and (potentially) loose the ordering you originally had. As a result, we prefer to allow duplicate regions (for now). I'd be happy to revisit this, though.